java-checks/src/main/java/org/sonar/java/checks/security/PubliclyWritableDirectoriesCheck.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.security;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ExpressionsHelper;
+import org.sonar.java.collections.SetUtils;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import static org.sonar.java.checks.helpers.ExpressionsHelper.getInvokedSymbol;
+
+@Rule(key = "S5443")
+public class PubliclyWritableDirectoriesCheck extends IssuableSubscriptionVisitor {
+
+  private static final String STRING_TYPE = "java.lang.String";
+  private static final String MESSAGE = "Make sure publicly writable directories are used safely here.";
+  private static final List<String> PUBLIC_WRITABLE_DIRS = Arrays.asList(
+    "/tmp",
+    "/var/tmp",
+    "/usr/tmp",
+    "/dev/shm",
+    "/dev/mqueue",
+    "/run/lock",
+    "/var/run/lock",
+    "/Library/Caches",
+    "/Users/Shared",
+    "/private/tmp",
+    "/private/var/tmp",
+    "\\Windows\\Temp",
+    "\\Temp",
+    "\\TMP");
+
+  private static final Set<String> TMP_DIR_ENV = SetUtils.immutableSetOf("TMP", "TMPDIR");
+
+  private static final MethodMatchers CREATE_FILE_MATCHERS = MethodMatchers.or(
```

Comment:
This matcher can match both constructors and method invocations and is only used when we already know the kind. This means that we can already say that half of the matcher will always fail when you are using it. You should split it in two.